### PR TITLE
Add Enums support to `substring` function

### LINKED ISSUE #57151: MySQL compatibility: substring with enums
Required by Tableau via MySQL interface.

Sample generated query:

```sql
SELECT 
  CONCAT(
    SUBSTRING(`cell_towers`.`radio`, 1, 1024), 
    SUBSTRING(`cell_towers`.`radio`, 1, 1024)
  ) AS `Calculation_3276650229452562433`
FROM `cell_towers`
GROUP BY 1
```

Simplified:

```sql
SELECT substring(CAST('foo', 'Enum(\'foo\' = 1)'), 1, 1000)
```

fails with 

```
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal type Enum8('foo' = 1) of argument of function substring: While processing substring(CAST('foo', 'Enum(\'foo\' = 1)'), 1, 1000). (ILLEGAL_TYPE_OF_ARGUMENT)
```

**Expected behavior**

Treat enums as strings when calling the `substring` function.

**How to reproduce**
* Which ClickHouse server version to use: latest

### PR DESCRIPTION
**Pull request by slvrtrn**
Resolves #57151 

Introduces `EnumSource` alongside `StringSource` and others, which allows usage of Enum8/Enum16 types in the `substring` function.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Function `substring()` (aliases: `substr`, `mid`) can now be used with `Enum` types. Previously, the first function argument had to be a value of type `String` or `FixedString`. This improves compatibility with 3rd party tools such as Tableau via MySQL interface.

### Documentation entry for user-facing changes
- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 497e5abc4793a8c486192593fc48c0b67db4d483 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57277/497e5abc4793a8c486192593fc48c0b67db4d483/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7011783682">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57277/497e5abc4793a8c486192593fc48c0b67db4d483/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by UnamedRus** [comment]
Does it return string or Enum as result?

Ie convert `Enum('hello' = -5, 'world' = 15)` to `Enum('he' = -5, 'wo' = 15)`

**Comment by slvrtrn** [comment]
@UnamedRus, String:

```
DESCRIBE TABLE
(
    SELECT substring(CAST('foo', 'Enum(\'foo\' = 1)'), 1, 1000)
)

┌─name─────────────────────────────────────────────────┬─type───┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ substring(CAST('foo', 'Enum(\'foo\' = 1)'), 1, 1000) │ String │              │                    │         │                  │                │
└──────────────────────────────────────────────────────┴────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

````

**Comment by UnamedRus** [comment]
It would be nice to have enum, so in most cases, you would need to only change mapping, but keep column values intact.

For case of clash, you would need to remap column values to some common id, but its still cheaper than operating on strings.

Another, probably simpler option is to return Lowcardinality(String) instead of String.

File: tests/queries/0_stateless/00493_substring_of_enum.sql [hunk 1]
```diff
@@ -0,0 +1,32 @@
+DROP TABLE IF EXISTS tab;
+CREATE TABLE tab(e8 Enum8('hello' = -5, 'world' = 15), e16 Enum16('shark' = -999, 'eagle' = 9999)) ENGINE MergeTree ORDER BY tuple();
+INSERT INTO TABLE tab VALUES ('hello', 'shark'), ('world', 'eagle');
+
```
Comment (rschu1ze) [inline tests/queries/0_stateless/00493_substring_of_enum.sql:4]: Such tests are smart but 1. they are difficult to debug in case something goes wrong and 2. they test unrelated functionality (joins) which can potentially break the test if something is wrong with joins.

My personal preference are much more "dumber tests", basically:
- check that the preconditions are checked (see `getReturnTypeImpl()`, e.g. one could test that `substringUTF8` with enum argument throws)
- then make a testing matrix of
1. const vs. non-const inputs (use materialize() or CREATE TABLE)
2. offset < 0 vs. offset == 0 vs. offset > 0 3. length < 0, length == 0, length > 0

But this is not a blocker, if you like we can keep the existing test.

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
> It would be nice to have enum, so in most cases, you would need to only change mapping, but keep column values intact.

> For case of clash, you would need to remap column values to some common id, but its still cheaper than operating on strings.

> Another, probably simpler option is to return Lowcardinality(String) instead of String.

This is a valid thought of course, but the main reason for this PR is to increase MySQL compatibility. In my view, running `substring()` on enums is a weird edge case and rarely useful in practice but certain 3rd party BI tools which expect MySQL behavior do it regardless. So we better implement MySQL behavior.

File: tests/queries/0_stateless/00493_substring_of_enum.sql (hunk 1 above)
Comment (slvrtrn) [reply tests/queries/0_stateless/00493_substring_of_enum.sql:4]: I will make them dumber :) Thanks.

File: src/Functions/substring.cpp [hunk 2]
```diff
@@ -31,40 +31,40 @@ namespace
...
-        return std::make_shared<FunctionSubstring>();
-    }
-
-    String getName() const override
-    {
-        return name;
-    }
 
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionSubstring>(); }
+    String getName() const override { return name; }
```
Comment (slvrtrn) [inline src/Functions/substring.cpp:41]: here is it a one-liner
```C++
String getName() const override { return name; }
```
however, you changed it in EnumSource like:
```C++
size_t rowNum() const
{
    return row_num;
}
```
?

Comment (rschu1ze) [reply src/Functions/substring.cpp:41]: I changed it in EnumSource to achieve consistent formatting with StringSource which is in the same file (Sources.h).

For functions, the one-liner approach is okay.

**Comment by rschu1ze** [comment]
There was a AST asan failure by one of the queries in the test:
https://s3.amazonaws.com/clickhouse-test-reports/57277/8eff62a4de53d367997bfb2f94b32378a356ea29/fuzzer_astfuzzerasan/report.html
It does not look directly related to substring() but I am afraid that it will pop up in other PRs when we merge it (and at that point, the PR will be reverted).
You may either 1. merge from master again (maybe the problem was a one-hit wonder but it does not look so) or, 2. try to reproduce the issue locally (using the query in above link), or 3. rewrite the tests to not use joins.

**Comment by slvrtrn** [comment]
@vdimir, thanks for a quick fix! Does your PR also cover [the errors from server.log in this one](https://s3.amazonaws.com/clickhouse-test-reports/57277/e275080f521150ad59f31b44c5201bdf0b6e2d90/fuzzer_astfuzzerubsan/report.html)?

File: tests/queries/0_stateless/00493_substring_of_enum.sql (hunk 1 above)
Comment (rschu1ze) [reply tests/queries/0_stateless/00493_substring_of_enum.sql:4]: https://github.com/ClickHouse/ClickHouse/pull/57821
